autoware_lanelet2_map_validator/src/validators/intersection/right_of_way_without_traffic_lights.cpp

```diff
@@ -0,0 +1,236 @@
+      continue;
+    }
+
+    auto virtual_traffic_light_elems =
+      lanelet.regulatoryElementsAs<lanelet::autoware::VirtualTrafficLight>();
+    if (!virtual_traffic_light_elems.empty()) {
+      continue;
+    }
+
+    const auto right_of_way_elems = lanelet.regulatoryElementsAs<lanelet::RightOfWay>();
+
+    // Issue-001: Lanelet with turn_direction but no traffic lights missing right_of_way reference
+    if (right_of_way_elems.empty()) {
+      std::map<std::string, std::string> reason_map;
+      reason_map["turn_direction"] = lanelet.attribute("turn_direction").value();
+      issues.emplace_back(
+        construct_issue_from_code(issue_code(this->name(), 1), lanelet.id(), reason_map));
+      continue;
+    }
+
+    // Process each right_of_way regulatory element (multiple are allowed)
+    for (const auto & right_of_way_elem : right_of_way_elems) {
+      auto right_of_way_lanelets =
+        right_of_way_elem->getParameters<lanelet::ConstLanelet>(lanelet::RoleName::RightOfWay);
+
+      // Issue-003: right_of_way regulatory element should have exactly one right_of_way role
+      if (right_of_way_lanelets.size() != 1) {
+        issues.emplace_back(
+          construct_issue_from_code(issue_code(this->name(), 3), right_of_way_elem->id()));
+        continue;
+      }
+
+      bool is_set_as_right_of_way = false;
+      for (const auto & param : right_of_way_lanelets) {
+        if (param.id() == lanelet.id()) {
+          is_set_as_right_of_way = true;
+          break;
+        }
+      }
+
+      // Issue-004: right_of_way regulatory element doesn't set this lanelet as right_of_way role
+      if (!is_set_as_right_of_way) {
+        issues.emplace_back(
+          construct_issue_from_code(issue_code(this->name(), 4), right_of_way_elem->id()));
+        continue;
+      }
+
+      auto yield_lanelets =
+        right_of_way_elem->getParameters<lanelet::ConstLanelet>(lanelet::RoleName::Yield);
+
+      std::vector<lanelet::ConstLanelet> conflicting_lanelets;
+      for (const auto & other_lanelet : map.laneletLayer) {
+        if (other_lanelet.id() == lanelet.id()) {
+          continue;
+        }
+
+        const auto relation = routing_graph->routingRelation(lanelet, other_lanelet);
+        if (relation == lanelet::routing::RelationType::Conflicting) {
+          bool has_same_source = false;
+
+          auto prev_lanelets_current = routing_graph->previous(lanelet);
+          auto prev_lanelets_other = routing_graph->previous(other_lanelet);
+
+          for (const auto & prev_current : prev_lanelets_current) {
+            for (const auto & prev_other : prev_lanelets_other) {
+              if (prev_current.id() == prev_other.id()) {
+                has_same_source = true;
+                break;
+              }
+            }
+            if (has_same_source) break;
+          }
+
+          if (!has_same_source) {
+            conflicting_lanelets.push_back(other_lanelet);
+          }
```

Comment:
can you define a lambda or utility function like `has_same_source(lane1, lane2)` ?